pkg/pods/manifest.go

```diff
@@ -29,33 +28,141 @@ type LaunchableStanza struct {
+	SetConfig(config map[interface{}]interface{})
+	SetRunAsUser(user string)
+	SetStatusPort(port int)
+	SetStatusHTTP(statusHTTP bool)
+	AddLaunchableStanza(launchableId string, stanza LaunchableStanza)
+}
+
+var _ ManifestBuilder = manifestBuilder{}
+
+func NewManifestBuilder() ManifestBuilder {
+	return manifestBuilder{&manifest{}}
+}
+
+func (m manifestBuilder) GetManifest() Manifest {
+	return m.manifest
+}
+
+type manifestBuilder struct {
+	*manifest
+}
+
+// Manifest contains the functionality in common between SignedManifest and
+// UnsignedManifest. It is immutable (read-only) and does not expose signature
+// information
+type Manifest interface {
+	ID() string
+	RunAsUser() string
+	Write(out io.Writer) error
+	ConfigFileName() (string, error)
+	WriteConfig(out io.Writer) error
+	PlatformConfigFileName() (string, error)
+	WritePlatformConfig(out io.Writer) error
+	GetLaunchableStanzas() map[string]LaunchableStanza
+	GetConfig() map[interface{}]interface{}
+	SHA() (string, error)
+	GetStatusPort() int
+	GetStatusHTTP() bool
+	Marshal() ([]byte, error)
+	SignatureData() (plaintext, signature []byte)
+
+	GetBuilder() ManifestBuilder
+}
+
+// assert manifest implements Manifest and UnsignedManifest
+var _ Manifest = &manifest{}
+
+type manifest struct {
 	Id                string                      `yaml:"id"` // public for yaml marshaling access. Use ID() instead.
 	RunAs             string                      `yaml:"run_as,omitempty"`
 	LaunchableStanzas map[string]LaunchableStanza `yaml:"launchables"`
 	Config            map[interface{}]interface{} `yaml:"config"`
 	StatusPort        int                         `yaml:"status_port,omitempty"`
 	StatusHTTP        bool                        `yaml:"status_http,omitempty"`
-	// these fields are required to track the original text if it was signed
-	raw       []byte
+
+	// Used to track the original bytes so that we don't reorder them when
+	// doing a yaml.Unmarshal and a yaml.Marshal in succession
+	raw []byte
+
+	// Signature related fields, may be empty if manifest is not signed
 	plaintext []byte
 	signature []byte
 }
 
-func (manifest *Manifest) ID() string {
+func (m *manifest) GetBuilder() ManifestBuilder {
+	builder := manifestBuilder{
+		&manifest{},
+	}
+	*builder.manifest = *m
+	builder.manifest.plaintext = []byte{}
+	builder.manifest.signature = []byte{}
+	builder.manifest.raw = []byte{}
+	return builder
+}
+
+func (manifest *manifest) ID() string {
 	return manifest.Id
 }
 
-func (manifest *Manifest) RunAsUser() string {
+func (m manifestBuilder) SetID(id string) {
+	m.manifest.Id = id
+}
+
+func (manifest *manifest) GetLaunchableStanzas() map[string]LaunchableStanza {
+	return manifest.LaunchableStanzas
+}
+
+func (manifest *manifest) AddLaunchableStanza(launchableId string, stanza LaunchableStanza) {
+	if manifest.LaunchableStanzas == nil {
+		manifest.LaunchableStanzas = make(map[string]LaunchableStanza)
+	}
+
+	manifest.LaunchableStanzas[launchableId] = stanza
```

Comment:
The builder only gets a shallow copy of the original data structure, so this map reference will alias with the original.  You should probably make a copy of the map in `GetBuilder()`.

Or, change the builder interface to have `SetLaunchables(...)` instead of `Add...`.